### ISSUE TITLE
Contextual error reporting for string matchers

### ISSUE THREAD

**Issue by GoogleCodeExporter** [etorreborre/specs#83]
```
I am writing Specs for XSLT transformation that produces text. I'm using
must_== matcher for comparing the result to expected string like this:

detailedDiffs("[]")
n3 must_== "<http://orsr.sk/KE/Sro/22939-V/2008-12-20> rdf:type "

where n3 is the result of my transformation.

The errors I encounter during implementation are usually extra spaces here
and there, sometimes extra newlines, and it is reported like this:
org.specs.runner.UserError: '<http://orsr.sk/KE/Sro/22939-V/2008-12-20>
rdf:type ' is not equal to '<http://orsr.sk/KE/Sro/22939-V/2008-12-20>
rdf:type [ ]'

I find this not very useful in my case, because I often compare large
multiline strings and have to search long lines for a difference between
the two strings.  Is there a way to display only some near context around
the difference, like is the case with JUnit's assertEquals? I was not able
to find it... Also the need to manually activate detailedDiffs gets
repetitive after few tests. 

For comparison, JUnit's assertEquals output makes it easier to locate the
error in my case:

org.specs.runner.UserError: expected:<...008-12-20> rdf:type []> but
was:<...008-12-20> rdf:type [ ]>
```

Original issue reported on code.google.com by `pavol.va...@gmail.com` on 19 Jul 2009 at 7:45

**Comment by GoogleCodeExporter** [comment]
```
This issue was closed by revision r1071.
```

Original comment by `etorrebo...@gmail.com` on 31 Jul 2009 at 5:18
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
In version 1.6.0, I still get suboptimal results:

org.specs.runner.UserError: ...2001/XMLSchema#>.

' is not equal to ...-ns#>.
@prefix rdfs:[X] <http://www.w3.org/...


compared to JUnit's assert:

org.specs.runner.UserError: expected:<...-ns#>.
@prefix rdfs:[X] <http://www.w3.org/...> but was:<...-ns#>.
@prefix rdfs:[] <http://www.w3.org/...>

See atached Spec for full example.
```

Original comment by `pavol.va...@gmail.com` on 14 Sep 2009 at 11:22

Attachments:
- [StringContextSpec.scala](https://storage.googleapis.com/google-code-attachments/specs/issue-83/comment-2/StringContextSpec.scala)

**Comment by GoogleCodeExporter** [comment]
Original comment by `etorrebo...@gmail.com` on 15 Sep 2009 at 12:40
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
There were actually several small issues in my code with the test case you 
submitted. 
I think I fixed them (in 1.6.1-SNAPSHOT), so the error message is now:

...-ns#>.
@prefix rdfs:[] <http://www.w3.org/... is not equal to ...-ns#>.
@prefix rdfs:[X] <http://www.w3.org/... (issue83.scala:19)

Note also that specs matchers should even be a bit better than junit's one in 
the 
case where you have several small differences:

val expected = """@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs:X <http://www.w3.org/2000/01/rdf-schema#>.
@prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>.
@prefix xs: <http://www.w3.org/2001/XMLSchema#>.
"""
    val actual = """@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>.
@prefix xs: <http://Ywww.w3.org/2001/XMLSchema#>.
"""

In that case (a "X" and a "Y" are different), the failure message is:

...-ns#>.
@prefix rdfs:[] <http://w...: <http://[Y]www.w3.org/2001/XMLS... is not equal 
to 
...-ns#>.
@prefix rdfs:[X] <http://w...: <http://[]www.w3.org/2001/XMLS... 
(issue83.scala:19)

so you can see the two differences clearly, whereas in JUnit it is:

expected:<...-ns#>.
    @prefix rdfs:[X <http://www.w3.org/2000/01/rdf-schema#>.
    @prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>.
    @prefix xs: <http://]www.w3.org/2001/XMLS...> but was:<...-ns#>.
    @prefix rdfs:[ <http://www.w3.org/2000/01/rdf-schema#>.
    @prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>.
    @prefix xs: <http://Y]www.w3.org/2001/XMLS...> 

Note also that there was a bug in the detailedDiffs. In specs-1.6.0 there are 
smart 
defaults which are triggered when the strings to compare are too long 
(http://code.google.com/p/specs/wiki/MatchersGuide#Smart_differences). Your 
case 
showed that this was not working properly. You should be able now to run the 
same 
spec again without the "detailedDiffs" declaration.

```

Original comment by `etorrebo...@gmail.com` on 15 Sep 2009 at 1:21
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
Original comment by `etorrebo...@gmail.com` on 14 Feb 2010 at 2:43
- Changed state: **Verified**
